app/src/tracing.h

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "constants.h"
+
+#include <ccf/base_endpoint_registry.h>
+#include <ccf/endpoint_context.h>
+#include <functional>
+#include <regex>
+#include <sstream>
+#include <time.h>
+
+namespace scitt
+{
+  constexpr std::string_view REQUEST_ID_HEADER = "x-ms-request-id";
+  constexpr std::string_view CLIENT_REQUEST_ID_HEADER =
+    "x-ms-client-request-id";
+
+  const std::regex CLIENT_REQUEST_ID_REGEX("^[0-9a-zA-Z-]+");
+
+  thread_local std::string request_id;
+  thread_local std::optional<std::string> client_request_id;
+
+  struct AppData
+  {
+    std::string request_id;
+    std::optional<std::string> client_request_id;
+  };
+
+  void clear_trace_state()
+  {
+    request_id = "";
+    client_request_id = std::nullopt;
+  }
+
+  int diff_timespec_ms(
```

Comment:
Should we have a higher resolution than ms? Or even better, use floating point seconds?

Because we are limited by CCF's 1ms ticking interval, we probably will never see anything between 0 and 1ms, but at least using a format with higher resolution will allow this to work if we get a better source of time in the future (maybe with SEV-SNP?).

Reply:
I don't think a higher resolution is useful here, but using floating point seconds may be worth considering if that's more standard. I'll have a look.

Reply:
I think I prefer keeping it as-is. If we really need a higher resolution we can always switch to floating ms later on, but I doubt it's useful to get that kind of resolution.